File: MTTFAdder.java
import java.util.concurrent.atomic.AtomicInteger;

public class MTTFAdder
{
	public static StringBuilder	stats	= new StringBuilder();

	public MTTFAdder(final int P, final int spares, final int active)
	{
		int cores = Runtime.getRuntime().availableProcessors();
		final int simulCount = (int) Math.round(80 / (double) cores);
		final double timeArr[][] = new double[cores][simulCount];
		final double timeSum[] = new double[cores];
		final double lambda = 0.05; // 1 / lambda = 20 days
		final int Q = 10; // ns
		if (P == 1)
			stats.append(lambda + "\t" + spares + "\t" + active + "\t");
		Thread workers[] = new Thread[cores];
		final AtomicInteger workerNum = new AtomicInteger(0);
		for (int j = 0; j < workers.length; j++) {
			workers[j] = new Thread(new Runnable() {
				@Override
				public void run()
				{
					int worker = workerNum.getAndIncrement();
					for (int i = 0; i < simulCount; i++) {
						Adder a =
							new Adder(active, spares, lambda, 86400.0, Q, Math.pow(10.0, -9.0));
						if (i == 0 && P == 1 && worker == 0)
							stats.append(a.timeStep + "\t" + AdderCore.k);
						while (!a.hasFailed(P));
						timeArr[worker][i] = a.getTimeOfDeath();
					}
					for (double time : timeArr[worker])
						timeSum[worker] += time;
				}
			});
			workers[j].start();
		}
		for (Thread worker : workers)
			try {
				worker.join();
			} catch (InterruptedException e) {
				e.printStackTrace();
			}
		double totalSum = 0;
		for (double timeTotal : timeSum)
			totalSum += timeTotal;
		System.out.println("Average time: " + (totalSum / (cores * simulCount)));
		stats.append("\t" + (totalSum / (cores * simulCount)));
	}

	public static void main(String[] args)
	{
		long start = System.currentTimeMillis();
		AdderCore.k = args.length > 2 ? Integer.parseInt(args[2]) : AdderCore.k;
		for (int i = 1; i < 4; i++)
			new MTTFAdder(i, args.length > 0 ? Integer.parseInt(args[0]) : 0, args.length > 1
				? Integer.parseInt(args[1]) : 1);
		System.out.println(stats);
		System.out.println("Time taken: "
			+ ((System.currentTimeMillis() - start) / Math.pow(10.0, 3)) + "s");
	}
}

File: mttf_adder.cpp
#include <cstdio>
#include <cstdlib>
#include <ctime>
#include <cmath>
#include <cstring>
#include <iostream>
#include <sstream>
#include <atomic>

#include <pthread.h>
#include <unistd.h>
#include <sys/time.h>
#include <sys/types.h>

#include "adder.h"
#include "adder_core.h"

typedef struct timeval timeval_t;

struct adder_args {
    int spares;
    int active;
    double lambda;
    int Q;
    double Qseconds;
    double lambdaSeconds;
    volatile double *timeSum;
    int P;
    int simul_count;
};

typedef struct adder_args adder_args_t;

using namespace std;

stringstream stats;
atomic_int worker_num(0);
atomic_int added(0);

void *mttf_worker(void *args_void)
{
    int worker = worker_num++;
    adder_args_t *args = (adder_args_t *)args_void;
    double timeArr[args->simul_count];
    for (int i = 0; i < args->simul_count; i++) {
        adder a(args->active, args->spares, args->lambda, args->lambdaSeconds,
            args->Q, args->Qseconds);
        if (added++ == 0)
            stats << '\t' << global_timeStep;
        while (!a.hasFailed(args->P));
        srand(time(NULL));
        timeArr[i] = a.getTimeOfDeath();
    }
    for (int i = 0; i < args->simul_count; i++)
        args->timeSum[worker] += timeArr[i];
    return NULL;
}

void mttf(int P)
{
    worker_num = 0;
    long cores = sysconf(_SC_NPROCESSORS_ONLN);
    long int simul_count = lround(80 / cores);
    volatile double timeSum[cores];
    for (int i = 0; i < cores; i++)
        timeSum[i] = 0;
    adder_args_t args;
    args.spares = 1;
    args.active = 1;
    args.lambda = 0.05;
    args.Q = 10;
    args.Qseconds = pow(10, -9);
    args.lambdaSeconds = 86400;
    args.P = P;
    args.simul_count = simul_count;
    args.timeSum = timeSum;
    if (P == 1)
        stats << args.lambda << '\t' << args.spares << '\t' << args.active;
    pthread_t threads[cores];
    for (int i = 0; i < cores; i++) {
        pthread_create(&threads[i], NULL, mttf_worker, (void *)&args);
    }
    for (int i = 0; i < cores; i++) {
        pthread_join(threads[i], NULL);
    }
    double total_sum = 0;
    for (int i = 0; i < cores; i++) {
        total_sum += timeSum[i];
    }
    double result = total_sum / (cores * simul_count);
    cout << "Average time: " << result << endl;
    stats << '\t' << result;
}

int main()
{
    timeval_t start, end;
    gettimeofday(&start, NULL);
    for (int i = 1; i <= 3; i++)
        mttf(i);
    cout << stats.str() << endl;
    gettimeofday(&end, NULL);
    printf("Time taken: %ld.%.6ld\n", (end.tv_sec - start.tv_sec),
        (long int)fdim(end.tv_usec, start.tv_usec));
    return 0;
}

